nimbus/src/lib.rs

```diff
@@ -51,68 +47,106 @@ impl NimbusClient {
         config: Option<RemoteSettingsConfig>,
         available_randomization_units: AvailableRandomizationUnits,
     ) -> Result<Self> {
-        let client = Client::new(&collection_name, config.clone())?;
-        let resp = client.get_experiments()?;
+        let http_client = Client::new(&collection_name, config)?;
         let db = Database::new(db_path)?;
         let nimbus_id = Self::get_or_create_nimbus_id(&db)?;
-        let enrolled_experiments =
-            evaluator::filter_enrolled(&nimbus_id, &available_randomization_units, &resp)?;
-        Ok(Self {
-            experiments: resp,
-            enrolled_experiments,
+        let instance = Self {
+            http_client,
+            available_randomization_units,
             app_context,
             db,
             nimbus_id,
-        })
+        };
+        // This is a little suspect but it's not clear what the right thing is.
+        // If we have no experiments we assume this is a first run so fetch
+        // them now.
+        if instance.get_all_experiments()?.is_empty() {
```

Comment:
We're gonna have to move this so we don't do I/O in the constructor, I wonder if it'll make it harder?

Reply:
We already do network and disk io in the constructor (disk io is opening the db). We do need to think about that, but I don't think this patch makes things any worse really.

Reply:
Because we're planning to get Nimbus vendored into mozilla-central before we do the work of sorting out the details of the IO threading stuff, I/O in the constructor will prevent us from landing in mozilla-central.  Which is to say, we need a very simple xcpshell test to land, and we've got one: it tries to construct an object, and right now that causes viaduct on desktop to assert and crash because it ends up trying to do IO on the main thread.  I don't know that this needs to necessarily block this patch from landing, but I do want to raise a flag here that this sort of I/O will start to block desktop vendoring soon.  CC @piatra

Reply:
Ok, I tweaked this to avoid the network IO in the ctor (but note that disk IO remains - we open the DB). Note also  that #38 might provide better general test strategy for desktop - you could configure the sdk with a file:// URL and we'd never do network IO - it's probably either that, or a test http server which presumably would *still* need to be backed by the file system, so we might as well cut out the middle-man :)